docs/custom_diagram.md

```diff
@@ -5,92 +5,40 @@
 
 # Custom Diagram
 
-`Custom diagram`s let's you create custom diagrams based on the data in the model. You define the data collection using a dictionary.
+`Custom diagram`s let's you create custom diagrams based on the data in the model. You define the data collection using an iterable and `Custom diagram` takes care of the rest.
 You can access `.custom_diagram` on any supported model element.
 
-## Collector definition
+??? example "Custom Diagram of `PP 1 `"
 
-### `get` and `include`
+    ``` py
+    import capellambse
 
-At every step of the collection, you can either `get` or `include` elements. `get` will simply get the element and `include` will include the element in the collection. `name` is the attribute name.
+    visited = set()
 
-```yaml
-get:
-    - name: inputs
-    include:
-        - name: exchanges
-        - name: links
-    - name: outputs
-    include:
-        - name: exchanges
-        - name: links
-    - name: ports
-    include:
-        - name: exchanges
-        - name: links
-```
+    def _collector(
+        target: m.ModelElement,
+    ) -> cabc.Iterator[m.ModelElement]:
+        if target.uuid in visited:
+            return
+        visited.add(target.uuid)
+        for link in target.links:
+            yield link
+            yield from _collector(link.source)
+            yield from _collector(link.target)
```

Comment:
Using a recursive function with persistent state as the only example isn't exactly what I'd describe as a gentle introduction into the topic. ;)

My main gripe with this is that the `visited` set never gets cleared. While it's technically fine for a one-shot script that renders exactly one diagram, it'll immediately start to break down if the script should render another diagram as well, because the collector function would see left-over garbage and randomly skip over objects that it was supposed to collect.

One relatively simple solution to this issue is adding a small wrapper around the "actual" collector, which creates a new (locally scoped) visited set, similar to this:

```python
def _collector(target):
    visited = set()
    def collect(target):
        ...
    yield from collect(target)
```